### ISSUE TITLE
OpenPGP: decryption of session key with RSA fails with custom cipher provider if the first byte is 0

### ISSUE THREAD

**Issue by lclement57** [bcgit/bc-java#1003]
Hello, we are using **bcpg-jdk15on v1.69** to read OpenPGP messages encrypted using RSA keys.

The private RSA key is stored in a Hardware Security Module, thus when we are performing the decryption of the session key we are not using **org.bouncycastle.jcajce.provider.asymmetric.rsa.CipherSpi**, but a class from a jar provided by our HSM provider.

We sometimes have an IllegalBlockSizeException in the Cipher engine when the library tries to decrypt the session key of the message from the Multiprecision Integer contained in Public-Key Encrypted Session Key Packet.

We found that this exception occurred everytime that when writing the PGP message, the ciphered session key had its first byte being a 0.

After digging in the library, we found that in the class JcePublicKeyDataDecryptorFactoryBuilder, it takes whatever the multiprecision integer from the public key packet is, and passes the bytes from it directly to the Cipher engine (after removing the first 2 bytes corresponding to the MPI length). See: https://github.com/bcgit/bc-java/blob/master/pg/src/main/java/org/bouncycastle/openpgp/operator/jcajce/JcePublicKeyDataDecryptorFactoryBuilder.java#L262

However, in a PGP message the multiprecision integer does not always have a size corresponding to the modulus of the RSA key (n), as all leading 0 bits are lost when converting the ciphertext to a multiprecision integer.

Thus when the first 8 bits of the ciphertext have a value of zero, the length of the MPI is no more (n + 2) bytes, but only (n - 1 + 2) bytes, meaning that when we send the MPI bytes without the first two bytes to the engine at decryption time, there are only (n - 1) bytes sent.

It means that we do not send the correct number of bytes to the Cipher engine to decrypt the key. The RSABlindedEngine that is used by Bouncy Castle works fine with it, but the RSA engine from our HSM provider throws an IllegalBlockSizeException.

According to the RFC 8017 which I think is the correct one (PKCS #1: RSA Cryptography Specifications Version 2.2), the way it is handled by the library from our HSM provider seems to be expected. In section 7.2.2 about RSAES-PKCS1-v1_5 decryption that given the input:

```
Input:

      K        recipient's RSA private key
      C        ciphertext to be decrypted, an octet string of length k,
               where k is the length in octets of the RSA modulus n
```

Then the first step of the decryption is: 
```
      1.  Length checking: If the length of the ciphertext C is not k
          octets (or if k < 11), output "decryption error" and stop.
```

Which seems to explain why the library from our HSM provider throws an IllegalBlockSizeException when we only send a ciphertext of (n - 1) bytes to it.

We fixed the issue by manually adding the "missing" 0 bytes in the Cipher engine in JcePublicKeyDataDecryptorFactoryBuilder#decryptSessionData when we detect them, before adding the bytes from the MPI (there can be more than one leading 0 bytes in ciphertext after encrypting the session key with RSA, that will absent in the MPI).

In conclusion, I think that the Bouncy Castle library should be doing this padding with 0 bytes, as the size of the MPI is not guaranteed to be equal to the size of the RSA key, and trying to decrypt a ciphertext of a size not equal to the modulus of the key should not be allowed. What do you think about it ?

**Comment by dghgit** [comment]
The code has been used with a few HSMs aready. I can grant that the data has to be appended with zeros readily enough, although it's a question as to who should. I think you'd only be able to do the check though if the PrivateKey implemented RSAPrivateKey (a lot don't) and was willing to return the modulus (some don't do this either). How do you recover the size of the modulus?

**Comment by lclement57** [comment]
We are using a Luna Network HSM, with the provided Luna Java Security Provider API ([Javadoc](https://thalesdocs.com/gphsm/luna/7/docs/network/Content/sdk/java/javadoc-JSP/10_3_0/index.html))

The private key is represented by a class [LunaPrivateKeyRsa](https://thalesdocs.com/gphsm/luna/7/docs/network/Content/sdk/java/javadoc-JSP/10_3_0/com/safenetinc/luna/provider/key/LunaPrivateKeyRsa.html) that indeed implements RSAPrivateKey and from which we retrieve the modulus.

I initially thought that we could retrieve this information from the Cipher class directly, by using the method `Cipher#getBlockSize()`. With the Bouncy Castle provider, it returns the expected size (512 for a 4096 bits key), but with the Luna Provider it returns 0. I am not sure about that, but the 0 from the Luna Provider seems to be correct as RSA should not be considered a block cipher from what I found online.

So for now, in our workaround, I just ended up adding an overload of the `JcePublicKeyDataDecryptorFactoryBuilder#build` method that takes the expected size of the byte array as a parameter. It looks like:

```
public PublicKeyDataDecryptorFactory build(final PrivateKey privKey, final int expectedDataSize)
{
    return new PublicKeyDataDecryptorFactory()
    {
        public byte[] recoverSessionData(int keyAlgorithm, byte[][] secKeyData) 
            throws PGPException 
        {
            if (keyAlgorithm == PublicKeyAlgorithmTags.ECDH) 
            {
                throw new PGPException("ECDH requires use of PGPPrivateKey for decryption");
            }
            return decryptSessionData(keyAlgorithm, privKey, secKeyData, expectedDataSize);
        }

        public PGPDataDecryptor createDataDecryptor(boolean withIntegrityPacket, int encAlgorithm, byte[] key) 
            throws PGPException 
        {
            return contentHelper.createDataDecryptor(withIntegrityPacket, encAlgorithm, key);
        }
    };
}


private byte[] decryptSessionData(int keyAlgorithm, PrivateKey privKey, byte[][] secKeyData, int expectedDataSize)
    throws PGPException 
{
   ...

     if (keyAlgorithm == PGPPublicKey.RSA_ENCRYPT
      || keyAlgorithm == PGPPublicKey.RSA_GENERAL)
  {
      byte[] bi = secKeyData[0];  // encoded MPI

      // Add leading 0 bytes not included in the MPI
      if (bi.length < expectedDataSize + 2) {
          int missingZeroBytes = expectedDataSize + 2 - bi.length;
          for (int i = 0; i < missingZeroBytes; i++) {
              c1.update(new byte[]{0});
          }
      }

      c1.update(bi, 2, bi.length - 2);
  }

   ...
}
``` 

And it is called using `new CustomJcePublicKeyDataDecryptorFactoryBuilder().build(privateKeyRsa, privateKeyRsa.getKeySize() / 8)`.

I am not confident enough with the library to integrate it in a better way.

**Comment by dghgit** [comment]
You said it throws an IllegalBlockSizeException if the block is too short? Is that correct?

**Comment by dghgit** [comment]
Second thought, if I'm correct checking the MPI to see if it was 32bit aligned and padding accordingly would also avoid this?

**Comment by lclement57** [comment]
If the block is too short, an IllegalBlockSizeException is thrown by the Luna Cipher engine:

```
Caused by: javax.crypto.IllegalBlockSizeException: Could not process input data: function 'C_Decrypt' returns 0x21
	at com.safenetinc.luna.provider.cipher.LunaCipher.engineDoFinal(LunaCipher.java:635)
	at com.safenetinc.luna.provider.cipher.LunaCipherRSAPKCS.engineDoFinal(LunaCipherRSAPKCS.java:194)
	at com.safenetinc.luna.provider.cipher.LunaCipher.engineDoFinal(LunaCipher.java:528)
	at java.base/javax.crypto.Cipher.doFinal(Cipher.java:2085)
	at org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder.decryptSessionData(JcePublicKeyDataDecryptorFactoryBuilder.java:303)
	... 13 common frames omitted
```

If this paragraph from the RFC 8017 applies: [7.2.1.  Encryption Operation](https://datatracker.ietf.org/doc/html/rfc8017#section-7.2.1)

```
RSAES-PKCS1-V1_5-ENCRYPT ((n, e), M)

   Input:

      (n, e)   recipient's RSA public key (k denotes the length in
               octets of the modulus n)
      M        message to be encrypted, an octet string of length
               mLen, where mLen <= k - 11

   Output:

      C        ciphertext, an octet string of length k
```

Which indicates that the ciphertext always has the length of the modulus, then yes, 32bit aligning and padding the MPI would safely "restore" the original value of the ciphertext, as the only reason leading to a shorter MPI is that the leading 0 are lost

**Comment by lclement57** [comment]
On second thought I'm not sure I understood 32bit aligning correctly, the modulus of the RSA key could be any value, eg 700 and not always a power of 2, would this solution work with that?

**Comment by dghgit** [comment]
Yes, it could be, but it generally isn't. Let me think about it a bit more though...

**Comment by dghgit** [comment]
I'm pretty sure this will never work for everything, but if I've understood the details above it should work better than before and at least for your provider (who should seriously consider doing the zero padding on doFinal before the RSA primitive is invoked...). Try 170b08 or later at https://www.bouncycastle.org/betas

**Comment by lclement57** [comment]
I confirm that the version 170b08 fixes the issue with our provider.
Thanks for your feedback, I will make sure to bring this suggestion to them.

**Comment by dghgit** [comment]
Thanks. I'll close this one off.
